modular_zubbers/code/modules/ratqueen/regalrat.dm

```diff
@@ -0,0 +1,163 @@
+// Adds rat fashion system
+/mob/living/basic/regal_rat
+	/// What kind of fashion are we rocking
+	var/datum/rat_fashion/current_look
+	/// Press this to change your rat outfit
+	var/datum/action/cooldown/rat_fashion/fashion_select
+
+/mob/living/basic/regal_rat/Initialize(mapload)
+	. = ..()
+	fashion_select = new(src)
+	fashion_select.Grant(src)
+	pick_random_look()
+
+/// Randomise how we look on init
+/mob/living/basic/regal_rat/proc/pick_random_look()
+	var/list/valid_starting_styles = list()
+	for (var/datum/rat_fashion/style_path as anything in subtypesof(/datum/rat_fashion))
+		if (!initial(style_path.allow_random))
+			continue
+		valid_starting_styles += new style_path()
+	if (!length(valid_starting_styles))
+		return
+	current_look = pick(valid_starting_styles)
+	current_look.apply(src)
+
+/mob/living/basic/regal_rat/death(gibbed)
+	current_look.on_death(src)
+	return ..()
+
+/mob/living/basic/regal_rat/revive(full_heal_flags, excess_healing, force_grab_ghost)
+	. = ..()
+	if(!.)
+		return
+	current_look.apply(src)
+
+/mob/living/basic/regal_rat/Destroy()
+	. = ..()
+	QDEL_NULL(fashion_select)
+	QDEL_NULL(current_look)
+
+/// Regal rat swallows sludge to transform themselves into a different looking rat
+/datum/action/cooldown/rat_fashion
+	name = "Rat King's Transformation"
+	desc = "Assume your true form, whatever you decide it is at the moment."
+	check_flags = AB_CHECK_CONSCIOUS
+	cooldown_time = 10 SECONDS
+	melee_cooldown_time = 0 SECONDS
+	button_icon = 'icons/mob/actions/actions_spells.dmi'
+	button_icon_state = "exit_possession"
+	background_icon_state = "bg_clock"
+	overlay_icon_state = "bg_clock_border"
+
+/datum/action/cooldown/rat_fashion/Activate(atom/target)
+	var/mob/living/basic/regal_rat/rat_owner = owner
+	if (!istype(rat_owner))
+		owner.balloon_alert(owner, "not a rat!")
+		qdel(src)
+		return
+
+	var/list/options = list()
+	var/list/picks_to_instances = list()
+	var/list/rat_styles = subtypesof(/datum/rat_fashion)
+	for (var/style_path as anything in rat_styles)
+		var/datum/rat_fashion/style = new style_path()
+		var/datum/radial_menu_choice/choice = style.get_radial_select()
+		options += list("[choice.name]" = choice)
+		picks_to_instances[choice.name] = style
+
+	var/pick = show_radial_menu(owner, owner, options, require_near = TRUE)
+	if (!pick)
+		return
+
+	var/datum/rat_fashion/chosen = picks_to_instances[pick]
+	if (rat_owner.current_look.name == chosen.name)
+		owner.balloon_alert(owner, "nothing to change!")
+		return
+
+	var/turf/origin = get_turf(owner)
+	owner.balloon_alert_to_viewers("shudders...")
+
+	if (!do_after(owner, 3 SECONDS, target = origin))
+		owner.balloon_alert(owner, "interrupted!")
+		return
+
+	rat_owner.current_look.remove(rat_owner)
+	rat_owner.current_look = chosen
+	rat_owner.current_look.apply(owner)
+	var/obj/effect/particle_effect/fluid/smoke/poof = new(origin)
+	poof.lifetime = 2 SECONDS
+	poof.color = "#5f5940"
```

Comment:
I'm surprised setting smoke parameters after creating it like this works. But there's other code that does this. Some code also uses /datum/effect_system/fluid_spread/smoke for what it's worth, which seems to be what most places use? Whatever the case, this is fine.